File: #95_Unique_Binary_Search_Trees_II/Solution.java
import java.util.*;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}

public class Solution {
    public List<TreeNode> generateTrees(int n) {
    	return generateTrees(1, n);
    }

    private List<TreeNode> generateTrees(int leftbound, int rightbound) {
    	List<TreeNode> result = new ArrayList<TreeNode>();
    	if (leftbound > rightbound) {
    		result.add(null); // this is the trick!
    		return result;
    	}
    	for (int i = leftbound; i <= rightbound; i++) {
	    	List<TreeNode> left = generateTrees(leftbound, i - 1);
	    	List<TreeNode> right = generateTrees(i + 1, rightbound);
	    	for (TreeNode lnd : left) {
	    		for (TreeNode rnd : right) {
	    			if (lnd == null) System.out.println("left is null");
	    			TreeNode root = new TreeNode(i);
	    			root.left = lnd;
	    			root.right = rnd;
	    			result.add(root);
	    		}
	    	}
    	}
    	return result;
    }

    public static void main(String[] args) {
    	Solution sol = new Solution();
    	List<TreeNode> result = sol.generateTrees(3);
    }
}

File: #95_Unique_Binary_Search_Trees_II/Solution.py
class TreeNode:
	def __init__(self, x):
		self.val = x
		self.left = None
		self.right = None

class Solution:
	# @param {integer} n
	# @return {TreeNode[]}
	def generateTrees(self, n):
		return self.generate([x for x in range(1, n + 1)])

	def generate(self, nums):
		if not nums:
			return [None]
		re = []
		for x in range(len(nums)):
			leftNodes = self.generate(nums[:x])
			rightNodes = self.generate(nums[x + 1:])
			for l in leftNodes:
				for r in rightNodes:
					node = TreeNode(nums[x])
					node.left = l
					node.right = r
					re.append(node)
		return re